docs/source/general/data.rst

```diff
@@ -2,11 +2,371 @@
+
+.. note::
+
+    [WIP] We are currently working on a new feature to convert fully implemented :class:`~flash.data.process.Preprocess`
+
+    and :class:`~flash.data.process.Postprocess` are fully implemented, to be fully deployable for ``Endpoints``
+
+    or ``BatchTransformJob`` directly from checkpoints. Stay tuned !
+
+*************************************
+How to customize existing datamodules
+*************************************
+
+Currently, Flash Tasks are implementing using :class:`~flash.data.process.Preprocess` and :class:`~flash.data.process.Postprocess`.
+However, it is not a hard requirement and one can still use :class:`~torch.data.utils.Dataset`, but we highly recommend
+using :class:`~flash.data.process.Preprocess` and :class:`~flash.data.process.Postprocess` instead.
+
+Example::
+
+    from flash.data.data_module import DataModule
+
+    dm = DataModule(train_dataset=MyDataset(train=True))
+    trainer = Trainer(fast_dev_run=True)
+    trainer.fit(model, data_module=dm)
+
+In order to customize Flash to your need, you need to know what are :class:`~flash.data.data_module.DataModule`
+and :class:`~flash.data.process.Preprocess` responsibilities.
+
+.. note::
+
+    From this point, we strongly encourage the readers to quickly check the :class:`~flash.data.process.Preprocess` API before getting further.
+
+The :class:`~flash.data.data_module.DataModule` provides ``classmethod`` helpers to build
+:class:`~flash.data.process.Preprocess` and :class:`~flash.data.data_pipeline.DataPipeline`,
+generate ``AutoDataset`` and populate DataLoaders from them.
+
+The :class:`~flash.data.process.Preprocess` contains the processing logic related to a given task. One can easily override hooks
+to customize a built-in :class:`~flash.data.process.Preprocess` for his need.
+
+Example::
+
+    from flash.vision import ImageClassificationData, ImageClassifier, ImageClassificationPreprocess
+
+    class CustomImageClassificationPreprocess(ImageClassificationPreprocess):
+
+        @staticmethod
+        def load_sample(sample) -> Tuple[Image.Image, int]:
+            # By default, ``ImageClassificationPreprocess`` expects ``.PNG`` or ``.JPB`` to be loaded into PIL Image.
+            # Assuming you have numpy image, just override this hook and add your own logic.
+            numpy_image_path, label = sample
+            return np.load(numpy_image_path), sample
+
+    datamodule = ImageClassificationData.from_folders(
+        train_folder="data/hymenoptera_data/train/",
+        val_folder="data/hymenoptera_data/val/",
+        test_folder="data/hymenoptera_data/test/",
+        preprocess_cls=CustomImageClassificationPreprocess
+    )
+
+
+*****************************************************
+How to build a Datamodule + Preprocess for a new task
+*****************************************************
+
+In this section, we will create a very simple ``ImageClassificationPreprocess`` with a ``ImageClassificationDataModule``.
```

Comment:
The example below shows a very simple ``ImageClassificationPreprocess`` with a ``ImageClassificationDataModule``.

Reply:
Done !